File: ms/core/sim-engine.js
// sim-engine.js - headless port of the C game's board generation and clicks.
//
// Implements, bit-for-bit, the board layout used by the real minesweeper.exe
// so simulated games and live games agree on the same seed:
//
//   * xorshift64 PRNG (minesweeper.c xorshift())
//   * place_mines(): pool of every cell outside the first click's 3x3, then a
//     partial Fisher-Yates draw using  k = rng() % n   while n shrinks; tiny
//     boards fall back to "only the clicked cell is safe".
//   * reveal_cell() flood fill with the auto-win / auto-flag-mines rules and
//     the auto-reveal-mines rule on loss.
//
// The board's click/flag/chord/board/state protocol matches the scripting CLI
// (minesweeper.c cli_*) so solver.js's play_game() drives it unmodified
// through the SimClient adapter.
//
// This is a 1:1 port of the Python sim_engine.py.  A single behavioral
// difference: 64-bit integers use BigInt (JS doubles cannot represent
// uint64), and the board arrays are Uint8Array instead of Python lists.

import path from "node:path";
import { fileURLToPath } from "node:url";

const MASK64 = (1n << 64n) - 1n;

// (rows, cols, mines) per difficulty, mirroring g_presets[].
export const PRESETS = {
  beginner: [8, 8, 10],
  intermediate: [16, 16, 40],
  expert: [16, 30, 99],
};

export const DEFAULTS = {
  beginner: "beginner",
  intermediate: "intermediate",
  expert: "expert",
};

// The game's xorshift64, advanced in place (one value per call).
export class Rng64 {
  constructor(seed) {
    this.s = BigInt(seed) & MASK64;
  }

  next() {
    let x = this.s;
    x ^= (x << 13n) & MASK64;
    x ^= x >> 7n;
    x ^= (x << 17n) & MASK64;
    x &= MASK64;
    this.s = x;
    return x;
  }
}

// Single-step convenience; state is a 1-element array (mutable holder).
export function xorshift(state) {
  const r = new Rng64(state[0]);
  const v = r.next();
  state[0] = r.s;
  return v;
}

function isBigInt(v) {
  return typeof v === "bigint";
}

// A tiny bit of conversion glue: every public entry accepts either a Number
// or a BigInt seed and normalises to BigInt (matching Python's `seed & MASK64`).
function toU64(seed) {
  return BigInt(seed) & MASK64;
}

export class SimBoard {
  constructor(marksEnabled = true) {
    this.marks_enabled = marksEnabled;
    this.rows = 0;
    this.cols = 0;
    this.mines = 0;
    this.difficulty = "beginner";
    this.rng = new Rng64(0n);
    this.seed = 0n;
    this.mine = [];
    this.adj = [];
    this.revealed = [];
    this.mark = [];
    this.opened = 0;
    this.started = 0;
    this.over = 0;
    this.flags = 0;
    this.time = 0;
    this.paused = 0;
  }

  _reset(rows, cols, mines, difficulty, seed) {
    this.rows = rows;
    this.cols = cols;
    this.mines = mines;
    this.difficulty = difficulty;
    this.seed = toU64(seed);
    this.rng = new Rng64(this.seed);
    const n = rows * cols;
    this.mine = new Uint8Array(n);
    this.adj = new Uint8Array(n);
    this.revealed = new Uint8Array(n);
    this.mark = new Uint8Array(n);
    this.opened = 0;
    this.started = 0;
    this.over = 0;
    this.flags = 0;
    this.time = 0;
    this.paused = 0;
  }

  // difficulty: 'beginner' | 'intermediate' | 'expert' | 'custom r c m'.
  new(difficulty, seed) {
    const parts = String(difficulty).split(/\s+/);
    const diff = parts[0].toLowerCase();
    if (diff === "custom" && parts.length >= 4) {
      const rows = parseInt(parts[1], 10);
      const cols = parseInt(parts[2], 10);
      const mines = parseInt(parts[3], 10);
      this._reset(rows, cols, mines, "custom", seed);
    } else if (Object.prototype.hasOwnProperty.call(PRESETS, diff)) {
      const [rows, cols, mines] = PRESETS[diff];
      this._reset(rows, cols, mines, diff, seed);
    } else {
      throw new Error("unknown difficulty: " + JSON.stringify(difficulty));
    }
    return "OK";
  }

  // Handle one CLI-style command line; returns the full reply text.
  command(line) {
    const toks = String(line).trim().split(/\s+/);
    if (toks.length === 0 || (toks.length === 1 && toks[0] === "")) {
      return "OK\nEND\n";
    }
    const cmd = toks[0].toLowerCase();
    try {
      if (cmd === "new") {
        if (toks.length >= 5 && toks[1].toLowerCase() === "custom") {
          this.new(
            `custom ${toks[2]} ${toks[3]} ${toks[4]}`,
            this.seed,
          );
        } else {
          this.new(toks.length > 1 ? toks[1] : "beginner", this.seed);
        }
        return "OK\nEND\n";
      }
      if (cmd === "click") {
        this.click(parseInt(toks[1], 10), parseInt(toks[2], 10));
        return "OK\nEND\n";
      }
      if (cmd === "flag") {
        this.flag(parseInt(toks[1], 10), parseInt(toks[2], 10));
        return "OK\nEND\n";
      }
      if (cmd === "chord") {
        this.chord(parseInt(toks[1], 10), parseInt(toks[2], 10));
        return "OK\nEND\n";
      }
      if (cmd === "refresh" || cmd === "ping") {
        return "OK\nEND\n";
      }
      if (cmd === "state") {
        return this._stateText();
      }
      if (cmd === "board") {
        return this.board().join("\n") + "\nEND\n";
      }
    } catch {
      return "ERR bad args\nEND\n";
    }
    return "ERR unknown command\nEND\n";
  }

  _stateText() {
    return [
      `difficulty=${this.difficulty}`,
      `rows=${this.rows}`,
      `cols=${this.cols}`,
      `mines=${this.mines}`,
      `flags=${this.flags}`,
      `opened=${this.opened}`,
      `time=${this.time}`,
      `started=${this.started}`,
      `over=${this.over}`,
      `paused=${this.paused}`,
      `marks=${this.marks_enabled ? 1 : 0}`,
      `seeded=1`,
      `seed=${this.seed}`,
      "END",
    ].join("\n") + "\n";
  }

  state() {
    const out = {};
    for (const ln of this._stateText().split("\n")) {
      if (ln === "END") continue;
      const eq = ln.indexOf("=");
      if (eq >= 0) out[ln.slice(0, eq)] = ln.slice(eq + 1);
    }
    return out;
  }

  _idx(r, c) {
    return r * this.cols + c;
  }

  _inb(r, c) {
    return r >= 0 && r < this.rows && c >= 0 && c < this.cols;
  }

  _computeAdj() {
    const { rows, cols } = this;
    for (let r = 0; r < rows; r++) {
      for (let c = 0; c < cols; c++) {
        let cnt = 0;
        for (let dr = -1; dr <= 1; dr++) {
          for (let dc = -1; dc <= 1; dc++) {
            if (dr === 0 && dc === 0) continue;
            const rr = r + dr;
            const cc = c + dc;
            if (this._inb(rr, cc) && this.mine[this._idx(rr, cc)]) cnt++;
          }
        }
        this.adj[this._idx(r, c)] = cnt;
      }
    }
  }

  _placeMines(sr, sc) {
    const pool = [];
    for (let r = 0; r < this.rows; r++) {
      for (let c = 0; c < this.cols; c++) {
        if (Math.abs(r - sr) <= 1 && Math.abs(c - sc) <= 1) continue;
        pool.push(this._idx(r, c));
      }
    }
    if (pool.length < this.mines) {
      // tiny board: only the clicked cell safe
      pool.length = 0;
      for (let r = 0; r < this.rows; r++) {
        for (let c = 0; c < this.cols; c++) {
          if (!(r === sr && c === sc)) pool.push(this._idx(r, c));
        }
      }
    }
    let n = pool.length;
    let placed = 0;
    while (placed < this.mines && n > 0) {
      const k = Number(this.rng.next() % BigInt(n));
      const idx = pool[k];
      pool[k] = pool[n - 1];
      n -= 1;
      if (!this.mine[idx]) {
        this.mine[idx] = 1;
        placed += 1;
      }
    }
    this._computeAdj();
  }

  _firstClick(r, c) {
    if (this.started) return 0;
    this.started = 1;
    this._placeMines(r, c);
    return 1;
  }

  _endGameLose() {
    if (this.over) return;
    this.over = -1;
    for (let i = 0; i < this.rows * this.cols; i++) {
      if (this.mine[i]) this.revealed[i] = 1;
    }
  }

  _endGameWin() {
    if (this.over) return;
    this.over = 1;
    for (let i = 0; i < this.rows * this.cols; i++) {
      if (this.mine[i] && this.mark[i] !== 1) {
        this.mark[i] = 1;
        this.flags += 1;
      }
    }
  }

  _revealCell(r, c) {
    if (!this._inb(r, c)) return;
    const i = this._idx(r, c);
    if (this.revealed[i] || this.mark[i] === 1) return;
    if (this.mine[i]) {
      this._endGameLose();
      return;
    }
    if (this.over) return;
    this.revealed[i] = 1;
    this.opened += 1;
    if (this.adj[i] === 0) {
      for (let dr = -1; dr <= 1; dr++) {
        for (let dc = -1; dc <= 1; dc++) {
          if (dr === 0 && dc === 0) continue;
          this._revealCell(r + dr, c + dc);
        }
      }
    }
    if (this.opened === this.rows * this.cols - this.mines) {
      this._endGameWin();
    }
  }

  click(r, c) {
    if (!this._inb(r, c)) return;
    if (!this.started) this._firstClick(r, c);
    this._revealCell(r, c);
  }

  _cycleMark(cell) {
    if (this.over) return;
    if (this.mark[cell] === 0) {
      this.mark[cell] = 1;
      this.flags += 1;
    } else if (this.mark[cell] === 1) {
      this.flags -= 1;
      this.mark[cell] = this.marks_enabled ? 2 : 0;
    } else {
      this.mark[cell] = 0;
    }
  }

  flag(r, c) {
    if (this._inb(r, c)) this._cycleMark(this._idx(r, c));
  }

  _doChord(cell) {
    const r = Math.floor(cell / this.cols);
    const c = cell % this.cols;
    let cnt = 0;
    for (let dr = -1; dr <= 1; dr++) {
      for (let dc = -1; dc <= 1; dc++) {
        if (dr === 0 && dc === 0) continue;
        const rr = r + dr;
        const cc = c + dc;
        if (this._inb(rr, cc) && this.mark[this._idx(rr, cc)] === 1) cnt++;
      }
    }
    if (cnt === this.adj[cell]) {
      for (let dr = -1; dr <= 1; dr++) {
        for (let dc = -1; dc <= 1; dc++) {
          if (dr === 0 && dc === 0) continue;
          const rr = r + dr;
          const cc = c + dc;
          if (this._inb(rr, cc)) this._revealCell(rr, cc);
        }
      }
    }
  }

  chord(r, c) {
    if (this._inb(r, c)) this._doChord(this._idx(r, c));
  }

  board() {
    const out = [];
    for (let r = 0; r < this.rows; r++) {
      let row = "";
      for (let c = 0; c < this.cols; c++) {
        const i = this._idx(r, c);
        if (!this.revealed[i]) {
          row += this.mark[i] === 1 ? "F" : this.mark[i] === 2 ? "?" : ".";
        } else if (this.mine[i]) {
          row += "*";
        } else {
          row += String.fromCharCode(48 + this.adj[i]);
        }
      }
      out.push(row);
    }
    return out;
  }
}

// Stand-in for a socket: sendall applies commands and queues replies.
export class SimSock {
  constructor(board) {
    this.board = board;
    this.queue = [];
  }

  sendall(data) {
    let text = data;
    if (typeof text !== "string") {
      text = Buffer.from(text).toString("ascii");
    }
    for (const rawLine of text.split("\n")) {
      const line = rawLine.trim();
      if (!line) continue;
      const reply = this.board.command(line);
      if (reply) {
        for (const l of reply.split("\n")) this.queue.push(l);
      }
    }
  }

  close() {}
}

// SimClient - solver.js-compatible adapter driving a SimBoard headlessly.
export class SimClient {
  constructor(sim = null, seed = 0, marksEnabled = true) {
    this.sim = sim || new SimBoard(marksEnabled);
    this.sim.seed = isBigInt(seed) ? seed : BigInt(seed);
    this.sock = new SimSock(this.sim);
  }

  _readLine() {
    if (this.sock.queue.length === 0) {
      throw new Error("no pending reply (protocol error)");
    }
    return this.sock.queue.shift();
  }

  cmd(text) {
    this.sock.sendall(text + "\n");
    const lines = [];
    while (true) {
      const line = this._readLine();
      if (line === "END") return lines;
      lines.push(line);
    }
  }

  new(difficulty) {
    return this.cmd("new " + difficulty);
  }

  click(r, c) {
    return this.cmd(`click ${r} ${c}`);
  }

  flag(r, c) {
    return this.cmd(`flag ${r} ${c}`);
  }

  chord(r, c) {
    return this.cmd(`chord ${r} ${c}`);
  }

  state() {
    return this.sim.state();
  }

  board() {
    return this.sim.board();
  }

  refresh(on) {
    return this.cmd(`refresh ${on ? 1 : 0}`);
  }

  close() {}
}

export function boardForSeed(difficulty, seed, click = null) {
  const b = new SimBoard();
  b.new(difficulty, seed);
  if (click !== null && click[0] !== null && click[1] !== null) {
    b.click(click[0], click[1]);
  }
  return b;
}

const isMain =
  process.argv[1] !== undefined &&
  path.resolve(process.argv[1]) === fileURLToPath(import.meta.url);
if (isMain) {
  // tiny self-check when run directly (mirrors sim_engine.py __main__)
  const b = boardForSeed("beginner", 12345, [3, 3]);
  const lost = b.over === -1;
  console.log(
    `rows=${b.rows} cols=${b.cols} mines=${b.mines} opened=${b.opened} over=${b.over} lost=${lost}`,
  );
  for (const ln of b.board()) console.log(ln);
  process.exit(0);
}

File: archive/server/sim_engine.py
"""sim_engine.py - headless port of the C game's board generation and clicks.

Implements, bit-for-bit, the board layout used by the real minesweeper.exe
so simulated games and live games agree on the same seed:

  * xorshift64 PRNG (minesweeper.c xorshift())
  * place_mines(): pool of every cell outside the first click's 3x3, then a
    partial Fisher-Yates draw using  k = rng() % n   while n shrinks; tiny
    boards fall back to "only the clicked cell is safe".
  * reveal_cell() flood fill with the auto-win / auto-flag-mines rules and
    the auto-reveal-mines rule on loss.

The board's click/flag/chord/board/state protocol matches the scripting CLI
(minesweeper.c cli_*) so ms_solver.py's play_game() drives it unmodified
through the SimClient adapter.
"""

import sys

MASK64 = (1 << 64) - 1

# A resolved seed of zero would leave xorshift64 stuck at zero forever (every
# output step is x ^= x<<13 ^ x>>7 ^ x<<17 with x == 0), so seed 0 is mapped
# onto a fixed nonzero constant.  This must stay identical in minesweeper.c
# (RNG_ZERO_SEED_FALLBACK) and ms/core/sim-engine.js.
ZERO_SEED_FALLBACK = 0x9E3779B97F4A7C15


def normalise_seed(seed):
    """Mask to uint64, mapping 0 onto ZERO_SEED_FALLBACK (see above)."""
    s = seed & MASK64
    return s if s else ZERO_SEED_FALLBACK

# (rows, cols, mines) per difficulty, mirroring g_presets[].
PRESETS = {
    "beginner": (8, 8, 10),
    "intermediate": (16, 16, 40),
    "expert": (16, 30, 99),
}

DEFAULTS = {"beginner": "beginner", "intermediate": "intermediate",
            "expert": "expert"}


class Rng64:
    """The game's xorshift64, advanced in place (one value per call)."""

    def __init__(self, seed):
        self.s = normalise_seed(seed)

    def next(self):
        x = self.s
        x ^= (x << 13) & MASK64
        x ^= x >> 7
        x ^= (x << 17) & MASK64
        x &= MASK64
        self.s = x
        return x


def xorshift(state):
    """Single-step convenience; state is a 1-element list (mutable holder)."""
    r = Rng64(state[0])
    v = r.next()
    state[0] = r.s
    return v


class SimBoard:
    """Headless game with the same semantics as the C Game struct."""

    def __init__(self, marks_enabled=True):
        self.marks_enabled = marks_enabled
        self.rows = 0
        self.cols = 0
        self.mines = 0
        self.difficulty = "beginner"
        self.rng = Rng64(0)
        self.seed = 0
        self.mine = []
        self.adj = []
        self.revealed = []
        self.mark = []
        self.opened = 0
        self.started = 0
        self.over = 0
        self.flags = 0
        self.time = 0
        self.paused = 0

    # ------------------------------------------------------------ lifecycle
    def _reset(self, rows, cols, mines, difficulty, seed):
        self.rows = rows
        self.cols = cols
        self.mines = mines
        self.difficulty = difficulty
        self.seed = normalise_seed(seed)
        self.rng = Rng64(self.seed)
        n = rows * cols
        self.mine = [False] * n
        self.adj = [0] * n
        self.revealed = [False] * n
        self.mark = [0] * n
        self.opened = 0
        self.started = 0
        self.over = 0
        self.flags = 0
        self.time = 0
        self.paused = 0

    def new(self, difficulty, seed):
        """difficulty: 'beginner' | 'intermediate' | 'expert' | 'custom r c m'."""
        parts = difficulty.split()
        diff = parts[0].lower()
        if diff == "custom" and len(parts) >= 4:
            rows = int(parts[1])
            cols = int(parts[2])
            mines = int(parts[3])
            self._reset(rows, cols, mines, "custom", seed)
        elif diff in PRESETS:
            rows, cols, mines = PRESETS[diff]
            self._reset(rows, cols, mines, diff, seed)
        else:
            raise ValueError("unknown difficulty: %r" % difficulty)
        return "OK"

    # ------------------------------------------------------------ protocol
    def command(self, line):
        """Handle one CLI-style command line; returns the full reply text."""
        toks = line.strip().split()
        if not toks:
            return "OK\nEND\n"
        cmd = toks[0].lower()
        try:
            if cmd == "new":
                if len(toks) >= 5 and toks[1].lower() == "custom":
                    self.new("custom %s %s %s" % (toks[2], toks[3], toks[4]),
                             self.seed)
                else:
                    self.new(toks[1] if len(toks) > 1 else "beginner", self.seed)
                return "OK\nEND\n"
            if cmd == "click":
                self.click(int(toks[1]), int(toks[2]))
                return "OK\nEND\n"
            if cmd == "flag":
                self.flag(int(toks[1]), int(toks[2]))
                return "OK\nEND\n"
            if cmd == "chord":
                self.chord(int(toks[1]), int(toks[2]))
                return "OK\nEND\n"
            if cmd == "refresh":
                return "OK\nEND\n"
            if cmd == "ping":
                return "OK\nEND\n"
            if cmd == "state":
                return self._state_text()
            if cmd == "board":
                return "\n".join(self.board()) + "\nEND\n"
        except (IndexError, ValueError):
            return "ERR bad args\nEND\n"
        return "ERR unknown command\nEND\n"

    def _state_text(self):
        return ("difficulty=%s\nrows=%d\ncols=%d\nmines=%d\nflags=%d\n"
                "opened=%d\ntime=%d\nstarted=%d\nover=%d\npaused=%d\n"
                "marks=%d\nseeded=1\nseed=%d\nEND\n" % (
                    self.difficulty, self.rows, self.cols, self.mines,
                    self.flags, self.opened, self.time, self.started,
                    self.over, self.paused, 1 if self.marks_enabled else 0,
                    self.seed))

    def state(self):
        out = {}
        for ln in self._state_text().splitlines():
            if ln == "END":
                continue
            if "=" in ln:
                k, v = ln.split("=", 1)
                out[k] = v
        return out

    # ------------------------------------------------------------ game core
    def _idx(self, r, c):
        return r * self.cols + c

    def _inb(self, r, c):
        return 0 <= r < self.rows and 0 <= c < self.cols

    def _compute_adj(self):
        for r in range(self.rows):
            for c in range(self.cols):
                cnt = 0
                for dr in (-1, 0, 1):
                    for dc in (-1, 0, 1):
                        if dr == 0 and dc == 0:
                            continue
                        rr, cc = r + dr, c + dc
                        if self._inb(rr, cc) and self.mine[self._idx(rr, cc)]:
                            cnt += 1
                self.adj[self._idx(r, c)] = cnt

    def _place_mines(self, sr, sc):
        pool = []
        for r in range(self.rows):
            for c in range(self.cols):
                if abs(r - sr) <= 1 and abs(c - sc) <= 1:
                    continue
                pool.append(self._idx(r, c))
        if len(pool) < self.mines:  # tiny board: only the clicked cell safe
            pool = [self._idx(r, c) for r in range(self.rows)
                    for c in range(self.cols) if not (r == sr and c == sc)]
        n = len(pool)
        placed = 0
        while placed < self.mines and n > 0:
            k = self.rng.next() % n
            idx = pool[k]
            pool[k] = pool[n - 1]
            n -= 1
            if not self.mine[idx]:
                self.mine[idx] = True
                placed += 1
        self._compute_adj()

    def _first_click(self, r, c):
        if self.started:
            return 0
        self.started = 1
        self._place_mines(r, c)
        return 1

    def _end_game_lose(self):
        if self.over:
            return
        self.over = -1
        for i in range(self.rows * self.cols):
            if self.mine[i]:
                self.revealed[i] = True

    def _end_game_win(self):
        if self.over:
            return
        self.over = 1
        for i in range(self.rows * self.cols):
            if self.mine[i] and self.mark[i] != 1:
                self.mark[i] = 1
                self.flags += 1

    def _reveal_cell(self, r, c):
        if not self._inb(r, c):
            return
        i = self._idx(r, c)
        if self.revealed[i] or self.mark[i] == 1:
            return
        if self.mine[i]:
            self._end_game_lose()
            return
        if self.over:
            return
        self.revealed[i] = True
        self.opened += 1
        if self.adj[i] == 0:
            for dr in (-1, 0, 1):
                for dc in (-1, 0, 1):
                    if dr == 0 and dc == 0:
                        continue
                    self._reveal_cell(r + dr, c + dc)
        if self.opened == self.rows * self.cols - self.mines:
            self._end_game_win()

    def click(self, r, c):
        if not self._inb(r, c):
            return
        if not self.started:
            self._first_click(r, c)
        self._reveal_cell(r, c)

    def _cycle_mark(self, cell):
        if self.over:
            return
        if self.mark[cell] == 0:
            self.mark[cell] = 1
            self.flags += 1
        elif self.mark[cell] == 1:
            self.flags -= 1
            self.mark[cell] = 2 if self.marks_enabled else 0
        else:
            self.mark[cell] = 0

    def flag(self, r, c):
        if self._inb(r, c):
            self._cycle_mark(self._idx(r, c))

    def _do_chord(self, cell):
        r = cell // self.cols
        c = cell % self.cols
        cnt = 0
        for dr in (-1, 0, 1):
            for dc in (-1, 0, 1):
                if dr == 0 and dc == 0:
                    continue
                rr, cc = r + dr, c + dc
                if self._inb(rr, cc) and self.mark[self._idx(rr, cc)] == 1:
                    cnt += 1
        if cnt == self.adj[cell]:
            for dr in (-1, 0, 1):
                for dc in (-1, 0, 1):
                    if dr == 0 and dc == 0:
                        continue
                    rr, cc = r + dr, c + dc
                    if self._inb(rr, cc):
                        self._reveal_cell(rr, cc)

    def chord(self, r, c):
        if self._inb(r, c):
            self._do_chord(self._idx(r, c))

    # ---------------------------------------------------------------- output
    def board(self):
        out = []
        for r in range(self.rows):
            row = []
            for c in range(self.cols):
                i = self._idx(r, c)
                if not self.revealed[i]:
                    row.append("F" if self.mark[i] == 1
                               else "?" if self.mark[i] == 2 else ".")
                elif self.mine[i]:
                    row.append("*")
                else:
                    row.append(chr(ord("0") + self.adj[i]))
            out.append("".join(row))
        return out


class SimSock:
    """Stand-in for a socket: sendall applies commands and queues replies."""

    def __init__(self, board):
        self.board = board
        self.queue = []

    def sendall(self, data):
        if isinstance(data, bytes):
            data = data.decode("ascii", "replace")
        for line in data.split("\n"):
            line = line.strip()
            if not line:
                continue
            reply = self.board.command(line)
            if reply:
                self.queue.extend(reply.splitlines())

    def close(self):
        pass


class SimClient:
    """MSClient-compatible adapter driving a SimBoard headlessly.

    play_game() in ms_solver.py talks to it exactly like a real client:
    new/click/flag/chord/state/board/refresh plus sock.sendall and
    _read_line() used by _batch().
    """

    def __init__(self, sim=None, seed=0, marks_enabled=True):
        self.sim = sim or SimBoard(marks_enabled=marks_enabled)
        self.sim.seed = seed
        self.sock = SimSock(self.sim)

    # ------------------------------------------------------------- MSClient
    def _read_line(self):
        if not self.sock.queue:
            raise ConnectionError("no pending reply (protocol error)")
        return self.sock.queue.pop(0)

    def cmd(self, text):
        self.sock.sendall(text + "\n")
        lines = []
        while True:
            line = self._read_line()
            if line == "END":
                return lines
            lines.append(line)

    def new(self, difficulty):
        return self.cmd("new " + difficulty)

    def click(self, r, c):
        return self.cmd("click %d %d" % (r, c))

    def flag(self, r, c):
        return self.cmd("flag %d %d" % (r, c))

    def chord(self, r, c):
        return self.cmd("chord %d %d" % (r, c))

    def state(self):
        return self.sim.state()

    def board(self):
        return self.sim.board()

    def refresh(self, on):
        return self.cmd("refresh %d" % (1 if on else 0))

    def close(self):
        pass


def board_for_seed(difficulty, seed, click=(None, None)):
    """Convenience: build a board for a seed and optionally apply a click."""
    b = SimBoard()
    b.new(difficulty, seed)
    if click != (None, None):
        b.click(*click)
    return b


if __name__ == "__main__":
    # tiny self-check when run directly
    b = board_for_seed("beginner", 12345, click=(3, 3))
    lost = b.over == -1
    print("rows=%d cols=%d mines=%d opened=%d over=%d lost=%s" %
          (b.rows, b.cols, b.mines, b.opened, b.over, lost))
    for ln in b.board():
        print(ln)
    sys.exit(0)
